docs/design.md

```diff
@@ -0,0 +1,432 @@
+   FullIndexRepo
+   ManifestRepo
+}
+
+type FullIndexRepo interface {
+   // public
+   GetFullIndex(key shard.Key) (FullIndex, error)
+
+   // private, called only by shard management when registering
+   // and destroying shards
+   InsertFullIndex(key shard.Key, index FullIndex) error
+   DeleteFullIndex(key shard.Key) (bool, error)
+}
+
+type ManifestRepo interface {
+   // public
+   GetManifest(key ManifestKey) (Manifest, error)
+   InsertManifest(key ManifestKey, manifest Manifest) error
+   DeleteManifest(key ManifestKey) (bool, error)
+   ExistsManifest(key ManifestKey) (bool, error)
+}
+
+type ManifestKey struct {
+   Shard    key.Shard
+   Rule     string
+   Version  string
+}
+
+type FullIndex interface {
+   Offset(c cid.Cid) (offset int64, err error) // inexistent: offset=-1, err=nil
+   Contains(c cid.Cid) (bool, error)
+   Len() (l int64, err error)
+   ForEach(func(c cid.Cid, offset int64) (ok bool, err error)) error
+}
+```
+
+## DAG access layer
+
+This layer is responsible for serving DAGs or sub-DAGs from one or many shards.
+Initially, queries will require the shard key. That is, queries will be bounded
+to a single identifiable shard.
+
+In the future, when the cross-shard top-level index is implemented, the DAG
+store will be capable of resolving the shard for any given root CID.
+
+DAG access layer allows various access abstractions:
+   1. Obtaining a Blockstore bound to a single shard.
+   2. Obtaining a Blockstore bound to a specific set of shards.
+   3. Obtaining a global Blockstore.
+   4. `ipld.Node` -- TBD.
+
+At this point, we are only concerned with (1). The remaining access patterns
+will be elaborated on in the future.
+
+```go
+type DAGAccessor interface {
+   Shard(key shard.Key) ShardAccessor
+}
+
+type ShardAccessor interface {
+   // Close() the accessor, signalling to the DAG store that we're done here and
+   // possible transient copies can be reclaimed once the shard is released.
+   io.Closer
+
+   Blockstore() (ReadBlockstore, error)
+}
+```
+
+## Requirements for CARv2
+
+- Index needs to be detachable.
+- Index offsets need to be relative to the CARv1, and not absolute in the
+  physical CARv2 file.
+- Index needs to be iterable.
+- Given any CAR file, we should be able to determine its version and
+  characteristics (concretely, indexed or not indexed).
+- Given a CARv1 file, it should be possible to generate a CARv2 index for it in
+  detached form.
+- Given a CARv2 file, we should be able to decompose it to the corresponding
+  CARv1 payload and the Carbs index. The CAR payload should be exposed with
+  `io.ReaderAt` and `io.Reader` abstractions.
+- Given an indexed CARv2 file, it should be possible to instantiate a
+  `ReadBlockstore` on it in a self-sufficient manner.
+- Given a CARv1 or unindexed CARv2, it should be possible to instantiate a
+  `ReadBlockstore` with an index provided by us.
+- It should be possible to write an indexed CARv2 in streaming fashion. The
```

Comment:
If this is strictly necessary then you'll have to know the CARv1 contents size before writing the header because that's included there along with the index offset. SelectiveCar I think had a method to "prepare" so you could, in theory, get this information with a first-pass prior to actual writing. Where possible, however, rewinding and updating the header would be much more efficient.

Reply:
The way we expect this to happen is that the index is written in a separate temp file and then concatenated with the output file on finalisation. That enables the output file to grow on the payload portion (CARv1) unbounded, thus supporting the streaming scenario.  On finalisation we go back and update the header with the final offsets/sizes of everything.

Reply:
noting that the carbon prototype will by default build up the index in memory - there's a hook where you can checkpoint it into a separate file as you need to, but because the index is getting insertion sorted on each new item, it isn't optimizing for recovery.

I think that's okay, but if you either want
(a) guarantees that on each append persisted to disk the index is also durable at the current contents of the partial car
or
(b) expect the index to not fit in memory
then there's a different write-ahead-log approach that should be used.

i think in reality this is fine and we just build up the index in memory during receive, and if we need to resume we re-build the partial index from the partial car.

Reply:
Is it OK to postpone this in the first pass or index in memory is a deal breaker for stream writes from get-go?